### PR TITLE
Return stable ids/nodes in ViolationLocation classes

### PR DESCRIPTION
**Pull request by olperr1**
**Please check if the PR fulfills these requirements**

- [X] The commit message follows our guidelines
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem?**

No

**What kind of change does this PR introduce?**

Bug fix

**What is the current behavior?**

#3153 introduces ViolationLocations for voltage violation.
- In the bus/breaker topology case, the bus id in the `BusBreakerViolationLocation` corresponds to a merged bus. This is not stable depending on the network implementation.
- In the node/breaker topology case, the `NodeBreakerViolationLocation` contains a list of busbar section ids. But in some cases there are no busbar section where the violation is detected.

**What is the new behavior (if this is a feature change)?**
- The `busId` of `BusBreakerViolationLocation` was replaced by a list of ids corresponding to the configured buses of the merge bus.
- The `busbarIds` of `NodeBreakerViolationLocation` was replaced by the list of the nodes where the violation was detected (always present).

Some unnecessary attributes / methods were also removed (for quality purpose):
- `getId()` (from `VoltageLocation` interface + both implementations)
- `getVoltageLevelId()` (from `VoltageLocation` + `BusBreakerViolationLocation`)
- `getBusId()` (from `VoltageLocation` interface + both implementations)
- `getBusbarIds()` (from `VoltageLocation` + both implementations)

**Does this PR introduce a breaking change or deprecate an API?**
- [X] Yes: There's breaking change compared to `powsybl:6.5.0-RC1`, **but not compared to any production release**.
- [ ] No

**If yes, please check if the following requirements are fulfilled**

- [X] The *Breaking Change* or *Deprecated* label has been added
- [ ] The migration steps are described in the following section

**What changes might users need to make in their application due to this PR? (migration steps)**

Only for migrating from 6.5.0-RC1:
- To retrieve data from a `ViolationLocation`, you now have to check it's type and cast it in the corresponding implementation.
- The `getBusId()` method from `BusBreakerViolationLocation` is replaced by a `getBusIds()` method returning the list of the ids of the **configured** buses instead of the sole id of the **merged** bus (which is not stable).
- The `getBusbarIds()` method from `NodeBreakerViolationLocation` is replaced by a `getNodes()` method returning the list of the nodes where the violation was detected. If needed, it is possible to retrieve the list of all concerned busbar sections using the following code:
```java
NodeBreakerViolationLocation vloc = (NodeBreakerViolationLocation) violationLocation;
VoltageLevel vl = network.getVoltageLevel(vloc.getVoltageLevelId());
List<String> busbarSectionIds = vloc.getNodes().stream()
        .map(node -> vl.getNodeBreakerView().getTerminal(node))
        .filter(Objects::nonNull)
        .map(Terminal::getConnectable)
        .filter(t -> t.getType() == IdentifiableType.BUSBAR_SECTION)
        .map(Identifiable::getId)
        .toList();
```

**Other information**:

The serialization versions of the short-circuits and security analysis results were increased in #3153.
The current version doesn't increase them since the said PR isn't already in a production release.

### CODE REVIEW

File: security-analysis/security-analysis-api/src/main/java/com/powsybl/security/NodeBreakerViolationLocation.java [hunk 1]
```diff
@@ -18,26 +18,18 @@ public class NodeBreakerViolationLocation implements ViolationLocation {
     private final List<String> busBarIds;
 
     public NodeBreakerViolationLocation(String voltageLevelId, List<String> busBarIds) {
-        Objects.requireNonNull(voltageLevelId, "voltageLevelId");
-        this.voltageLevelId = voltageLevelId;
-        this.busBarIds = busBarIds;
+        this.voltageLevelId = Objects.requireNonNull(voltageLevelId, "voltageLevelId should not be null");
```
Comment (EtienneLt) [inline security-analysis/security-analysis-api/src/main/java/com/powsybl/security/NodeBreakerViolationLocation.java:21]: do not we remove the voltageLevelId in Location ?

Comment (geofjamg) [reply security-analysis/security-analysis-api/src/main/java/com/powsybl/security/NodeBreakerViolationLocation.java:21]: We could as it is redundant we busbar section Ids which allows to get the voltage level directly BUT 
 - using the busbar section is a design issue too
 - we should use nodes (integer) instead to covers all use cases and be symmetrical with what is in the  BusBreakerViolationLocation

So maybe we should keep it and in a next PR, remove busbar section and replace by nodes.

File: security-analysis/security-analysis-api/src/main/java/com/powsybl/security/NodeBreakerViolationLocation.java [hunk 2]
```diff
@@ -18,26 +18,18 @@ public class NodeBreakerViolationLocation implements ViolationLocation {
     private final List<String> busBarIds;
```
Comment (geofjamg) [inline security-analysis/security-analysis-api/src/main/java/com/powsybl/security/NodeBreakerViolationLocation.java:18]: busbarSectionIds

File: security-analysis/security-analysis-api/src/main/java/com/powsybl/security/NodeBreakerViolationLocation.java (hunk 1 above)
Comment (geofjamg) [reply security-analysis/security-analysis-api/src/main/java/com/powsybl/security/NodeBreakerViolationLocation.java:21]: Or maybe we should do it now?

File: security-analysis/security-analysis-api/src/main/java/com/powsybl/security/BusBreakerViolationLocation.java [hunk 3]
```diff
@@ -7,35 +7,21 @@
...
 package com.powsybl.security;
 
+import java.util.List;
 import java.util.Objects;
-import java.util.Optional;
 
 /**
  * @author Étienne Lesot {@literal <etienne.lesot at rte-france.com>}
  */
 public class BusBreakerViolationLocation implements ViolationLocation {
```
Comment (geofjamg) [inline security-analysis/security-analysis-api/src/main/java/com/powsybl/security/BusBreakerViolationLocation.java:16]: We should write in the Java doc that it is Ids of configured buses of the bus/breaker view that are expected.

File: security-analysis/security-analysis-api/src/main/java/com/powsybl/security/NodeBreakerViolationLocation.java (hunk 1 above)
Comment (geofjamg) [reply security-analysis/security-analysis-api/src/main/java/com/powsybl/security/NodeBreakerViolationLocation.java:21]: To avoid a painful migration

**Review by geofjamg — APPROVED** [review]

File: security-analysis/security-analysis-api/src/main/java/com/powsybl/security/ViolationLocation.java [hunk 4]
```diff
@@ -21,18 +17,5 @@ enum Type {
...
-
-    default Optional<String> getBusId() {
-        return Optional.empty();
-    }
-
-    default List<String> getBusBarIds() {
-        return Collections.emptyList();
-    }
-
 }
```
Comment (vidaldid-rte) [inline security-analysis/security-analysis-api/src/main/java/com/powsybl/security/ViolationLocation.java:21]: Sans méthode, l'API sera dure à utiliser sans interface.

On pourrait ajouter getBusView(Network).getBusStream()
et getBusBreakerView(Network).getBusStream()

Comment (vidaldid-rte) [reply security-analysis/security-analysis-api/src/main/java/com/powsybl/security/ViolationLocation.java:34]: @olperr1 @geofjamg small addition to make API more user friendly
https://github.com/powsybl/powsybl-core/pull/3178/commits/aa833ab7c1f2ea35599f613d39c31f8fd7c872c9

Can you have a look ?

Comment (olperr1) [reply security-analysis/security-analysis-api/src/main/java/com/powsybl/security/ViolationLocation.java:34]: Thanks for your addition!

**Review by vidaldid-rte — APPROVED** [review]

**Pull request merged**